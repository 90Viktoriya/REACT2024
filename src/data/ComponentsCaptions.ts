export enum ComponentsCaptions {
  SEARCH = 'Search',
  ERROR = 'Catch Error',
  NOTHING_FOUND = 'Sorry, nothing found',
  ERROR_MSG = 'You throw the error',
  ERROR_TITLE = 'OOPS! Something went wrong',
  UNKNOWN_VALUE = 'unknown',
  ERROR_PAGE = 'OOPS! Page not found',
  PREV = 'Prev',
  NEXT = 'Next',
  CLOSE = 'Close'
}
