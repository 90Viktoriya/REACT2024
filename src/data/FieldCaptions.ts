export enum FieldCaptions {
  NAME = 'Name: ',
  GENDER = 'Gender: ',
  YEAR_OF_BIRTH = 'Year of birth: ',
  YEAR_OF_DEATH = 'Year of death: ',
  EPISODES = 'Episodes',
  MOVIES = 'Movies',
  PERFORMERS = 'Performers',
  DATE_OF_BIRTH = 'Date of birth: ',
  PLACE_OF_BIRTH = 'Place of birth',
  TITLE = 'Title: ',
  EPISODE_NUMBER = 'Episode number: ',
  RELEASE_DATE = 'Release date: '
}
