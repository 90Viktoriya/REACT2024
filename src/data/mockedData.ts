export const details = {
  uid: 'test12',
  name: 'Detailed card test',
  gender: '',
  yearOfBirth: 0,
  placeOfBirth: 'test',
  yearOfDeath: 0,
  performers: [{ uid: 'qwe', dateOfBirth: '2024', name: 'sd', placeOfBirth: 'aaa' }],
  episodes: [{ uid: 'asd', title: 'qqq', episodeNumber: 3 }],
  movies: [{ uid: 'zxc', title: 'qqq', usReleaseDate: '102' }]
};

export const page = {
  pageNumber: 1,
  pageSize: 1,
  numberOfElements: 1,
  totalElements: 1,
  totalPages: 1,
  firstPage: true,
  lastPage: false
};

export const characters = [
  {
    uid: 'ts',
    name: 'Character1',
    gender: 'M',
    yearOfBirth: 2024
  },
  {
    uid: 'ts2',
    name: 'aaa2',
    gender: 'F',
    yearOfBirth: 2024
  }
];

export const data = {
  page,
  characters
};
