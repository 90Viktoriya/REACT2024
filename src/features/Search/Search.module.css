.btn {
  margin-left: 0.1em;
  font-size: 1em;
  cursor: pointer;
  background-color: var(--brown);
  border: 2px outset var(--dark);
  color: var(--purple);
  border-radius: 8px;
  width: 100px;
  height: 40px;
}
.input {
  font-size: 1.5em;
  color: var(--dark);
  background-color: var(--light);
}