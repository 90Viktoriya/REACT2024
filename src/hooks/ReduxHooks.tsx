import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import type { RootState, AppDispatch } from '../store/storeRedux';

export const useAppDispatch = useDispatch.withTypes<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
//export const useAppSelector: RootState = useSelector.withTypes<RootState>();
