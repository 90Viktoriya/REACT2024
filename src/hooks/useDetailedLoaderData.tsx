import { useLoaderData } from 'react-router-dom';
import { detailedLoader } from '../features/Router/Loaders/detailedLoader';

export const useDetailedLoaderData = () => {
  return useLoaderData() as Awaited<ReturnType<typeof detailedLoader>>;
};
