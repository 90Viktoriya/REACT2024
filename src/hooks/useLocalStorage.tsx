import { useState, useEffect } from 'react';
import { Store } from '../store/store';

export function useLocalStorage() {
  const [searchValue, setSearchValue] = useState<string>(() => Store.getSearchValue());
  useEffect(() => {
    if (searchValue === undefined) {
      return;
    }
    Store.setSearchValue(searchValue);
  }, [searchValue]);
  return { searchValue, setSearchValue };
}
