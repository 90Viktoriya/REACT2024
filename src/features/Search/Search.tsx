import { MouseEventHandler, useCallback, useState } from 'react';
import { ComponentsCaptions } from '../../data/ComponentsCaptions';
import styles from './Search.module.css';
import { useFetcher, useLocation, useNavigate } from 'react-router-dom';
import { RouterPath } from '../Router/Router.enum';

export function Search({ searchValue }: { searchValue: string }) {
  const [inputValue, setInputValue] = useState(searchValue);
  const fetcher = useFetcher();

  const navigate = useNavigate();
  const location = useLocation();

  const handleOnClick: MouseEventHandler = useCallback(
    (event) => {
      if (
        event.currentTarget instanceof HTMLElement &&
        event.currentTarget === event.target &&
        location.pathname.includes(RouterPath.DETAILS)
      ) {
        navigate('..');
      }
    },
    [location.pathname, navigate]
  );
  const handleChange: React.ChangeEventHandler<HTMLInputElement> = (event) => {
    setInputValue(event?.target.value);
  };

  return (
    <fetcher.Form method="post" className={styles.search}>
      <input
        name="searchValue"
        className={styles.input}
        type="text"
        value={inputValue}
        onChange={handleChange}
        onClick={handleOnClick}
      />
      <button type="submit" className={styles.btn}>
        {ComponentsCaptions.SEARCH}
      </button>
    </fetcher.Form>
  );
}
