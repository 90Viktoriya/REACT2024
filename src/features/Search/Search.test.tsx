import { describe, expect, Mock, vi } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';
import { Search } from './Search';
import { useLocalStorage } from '../../hooks/useLocalStorage';
import { ComponentsCaptions } from '../../data/ComponentsCaptions';
import { MemoryRouter } from 'react-router-dom';

const mockedFunction = vi.fn();

vi.mock('react-router-dom', async () => {
  const mod = await vi.importActual('react-router-dom');
  return {
    ...mod,
    useNavigate: () => mockedFunction,
    useLocation: () => ({ pathname: '/details/123' })
  };
});

vi.mock('../../hooks/ReduxHooks', () => ({
  useAppDispatch: () => vi.fn()
}));

vi.mock('../../hooks/useLocalStorage', () => ({
  useLocalStorage: vi.fn()
}));

describe('Given Search component', async () => {
  it('when rendered, should match snapshot', () => {
    (useLocalStorage as Mock).mockReturnValue({
      searchValue: 'search'
    });
    const { asFragment } = render(<Search />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('When click button search, searchValue should changed', () => {
    const setValue = vi.fn();
    (useLocalStorage as Mock).mockReturnValue({
      searchValue: 'search',
      setValue
    });
    const screen = render(
      <MemoryRouter>
        <Search />
      </MemoryRouter>
    );

    const button = screen.queryByText(ComponentsCaptions.SEARCH) as HTMLElement;
    fireEvent.click(button);
    expect(setValue).toBeCalled();
    expect(mockedFunction).toBeCalled();
  });
  it('When click on input, details should be closed', () => {
    const setValue = vi.fn();
    (useLocalStorage as Mock).mockReturnValue({
      searchValue: 'search',
      setValue
    });
    const screen = render(
      <MemoryRouter>
        <Search />
      </MemoryRouter>
    );

    const input = screen.queryByRole('textbox') as HTMLElement;
    fireEvent.click(input);
    expect(mockedFunction).toBeCalled();
  });
  it('When input value changing, it should be changed', () => {
    render(
      <MemoryRouter>
        <Search />
      </MemoryRouter>
    );

    const input = screen.queryByRole('textbox') as HTMLInputElement;
    fireEvent.change(input, { target: { value: 'New Value' } });
    expect(input.value).toBe('New Value');
  });
});
