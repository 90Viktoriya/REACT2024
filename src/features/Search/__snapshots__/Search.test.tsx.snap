// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Given Search component > when rendered, should match snapshot 1`] = `
<DocumentFragment>
  <form
    class="_search_b6bc0a"
    method="post"
  >
    <input
      class="_input_b6bc0a"
      name="searchValue"
      type="text"
      value="Test Value"
    />
    <button
      class="_btn_b6bc0a"
      type="submit"
    >
      Search
    </button>
  </form>
</DocumentFragment>
`;
